design/upgrade.md

```diff
@@ -0,0 +1,179 @@
+
+If the upgrade controller uses the built-in rolling update support for certain stateless components, it should still verify all cluster health checks before proceeding with the next set of components.
+
+### **Synchronization**
+The upgrade process should be carefully orchestrated in a controlled manner to ensure reliability and success.
+Therefore, there should be some locking or synchronization that can ensure that while an upgrade is in progress, other changes to the cluster cannot be made.
+For example, if the upgrade controller is currently rolling out a new version, it should not be possible to modify the cluster CRD with other changes, such as removing a node from the cluster.
+This could be done by the operator stopping its watches on all CRDs or it could choose to simply return immediately from CRD events while the upgrade is in progress.
+
+There are also some mechanisms within Ceph that can help the upgrade proceed in a controlled manner.
+For example, the `noout` flag can be set in the Ceph cluster, indicating that while OSDs will be taken down to upgrade them, they should not be marked out of the cluster, which would trigger unnecessary recovery operations.
+The [Ceph Luminous upgrade guide](http://ceph.com/releases/v12-1-4-luminous-rc-released/#upgrading) recommends setting the `noout` flag for the duration of the upgrade.
+Details of the `noout` flag can be found in the [Ceph documentation](http://docs.ceph.com/docs/giant/rados/troubleshooting/troubleshooting-osd/#stopping-w-out-rebalancing).
+
+### **Scalability**
+For small clusters, the process of upgrading one pod at a time should be sufficient.
+However, for much later clusters (100+ nodes), this would result in an unacceptably long upgrade window duration.
+The upgrade controller should be able to batch some of its efforts to upgrade multiple pods at once in order to finish an upgrade in a more timely manner.
+
+This batching should not be done across component types (e.g. upgrading mons and OSDs at the same time), those boundaries where the health of the entire cluster is verified should still exist.
+This batching should also not be done for monitors as there are typically only a handful of monitors servicing the entire cluster and it is not recommended to have multiple monitors down at the same time.
+
+But, the upgrade controller should be able to slowly increase its component update batch size as it proceeds through some other component types, such as OSDs, MDS and RGW.
+For example, in true canary deployment fashion, a single OSD could be upgraded to the new version and OSD/cluster health will be verified.
+Then two OSDs could be updated at once and verification occurs again, followed by four OSDs, etc. up to a reasonable upper bound.
+We do not want too many pods going down at one time, which could potentially impact cluster health and functionality, so a sane upper bound will be important.
```

Comment:
@liewegas mentioned in another thread that I've lost that there is an osd field indicating it is safe to be upgraded. I'm not familiar with this, but it seems like it would be possible to collect some `N` number of osds that show that they can be upgraded and then do so assuming I haven't misinterpreted something.

Reply:
i remember mention of that flag as well, but we haven't been able to use it yet. Maybe at least we mention it in the upgrade doc so users know it is safe to upgrade. Note that currently the operator only cycles through one osd at a time and there is no parallelism.